packages/pigeon/tool/shared/generation.dart

```diff
@@ -0,0 +1,231 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:path/path.dart' as p;
+
+import 'process_utils.dart';
+
+enum GeneratorLanguages {
+  cpp,
+  java,
+  kotlin,
+  objc,
+  swift,
+}
+
+// A map of pigeons/ files to the languages that they can't yet be generated
+// for due to limitations of that generator.
+const Map<String, Set<GeneratorLanguages>> _unsupportedFiles =
+    <String, Set<GeneratorLanguages>>{
+  'enum_args': <GeneratorLanguages>{GeneratorLanguages.cpp},
+};
+
+String _snakeToPascalCase(String snake) {
+  final List<String> parts = snake.split('_');
+  return parts
+      .map((String part) =>
+          part.substring(0, 1).toUpperCase() + part.substring(1))
+      .join();
+}
+
+// Remaps some file names for Java output, since the filename on Java will be
+// the name of the generated top-level class. In some cases this is necessary
+// (e.g., "list", which collides with the Java List class in tests), and in
+// others it is just preserving previous behavior from the earlier Bash version
+// of the generation to minimize churn during the migration.
+// TODO(stuartmorgan): Remove the need for this when addressing
+// https://github.com/flutter/flutter/issues/115168.
+String _javaFilenameForName(String inputName) {
+  const Map<String, String> specialCases = <String, String>{
+    'android_unittests': 'Pigeon',
+    'host2flutter': 'Host2Flutter',
+    'list': 'PigeonList',
+    'message': 'MessagePigeon',
+    'voidflutter': 'VoidFlutter',
+    'voidhost': 'VoidHost',
+  };
+  return specialCases[inputName] ?? _snakeToPascalCase(inputName);
+}
+
+Future<int> generatePigeons({required String baseDir}) async {
+  // TODO(stuartmorgan): Make this dynamic rather than hard-coded. Or eliminate
+  // it entirely; see https://github.com/flutter/flutter/issues/115169.
+  const List<String> inputs = <String>[
+    'all_datatypes',
+    'all_void',
+    'android_unittests',
+    'async_handlers',
+    'background_platform_channels',
+    'enum_args',
+    'enum',
+    'host2flutter',
+    'java_double_host_api',
+    'list',
+    'message',
+    'multiple_arity',
+    'non_null_fields',
+    'null_fields',
+    'nullable_returns',
+    'primitive',
+    'void_arg_flutter',
+    'void_arg_host',
+    'voidflutter',
+    'voidhost',
+  ];
+
+  final String outputBase = p.join(baseDir, 'platform_tests', 'test_plugin');
+  final String alternateOutputBase =
+      p.join(baseDir, 'platform_tests', 'alternate_language_test_plugin');
+  // TODO(stuartmorgan): Eliminate this and use alternateOutputBase.
+  // See https://github.com/flutter/packages/pull/2816.
+  final String iosObjCBase =
+      p.join(baseDir, 'platform_tests', 'ios_unit_tests');
+
+  for (final String input in inputs) {
```

Comment:
It worth trying parallelizing this: https://pub.dev/packages/pmap  The input size is small and the execution time is large.

Reply:
I forgot to move your existing TODO about this; I've added it. (But also, https://github.com/flutter/flutter/issues/115168 will largely make this much less of an issue.)